### PR TITLE
Add better support for Arabic language to Text Layer

### LINKED ISSUE #507: RTL support for Text Layer
**Synfig version & platform**: 1.31, AppImage version on Debian(Sid)

**Issue description**: I had trouble implementing Arabic titles for one of my friends while doing a animated promotional clip. It was stalled, just because I know only Synfig for animation and Synfig didn't well with bi-directional language such as Arabic. (I could have done it with GIMP and imported it as static raster layer, but I wanted to morph every single letter transformed into a person :-()  Is there a possibility for implementing Complex Text Layout with the use of library called HarfBuzz? (I'm sorry if it has been already implemented) If it has been done already, could you tell me how to do it?

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/56368679-rtl-support-for-text-layer?utm_campaign=plugin&utm_content=tracker%2F321991&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F321991&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by rodolforg**
Arabic, Hebrew and other languages that needs text shaping is now finally supported, thanks to FriBiDi and HarfBuzz libraries.
These libraries are already backend for Gtk/Glib, so there is not any change for packaging.

Fix #507, #1621

### CODE REVIEW

**Comment by rodolforg** [comment]
Actually I don't know anything about those languages.
I tested only compared some random web texts appearance with that Synfig now provides.
So, please, test it – if you find any problem, and try to explain me what is wrong.

@Fahad-Alsaidi could you test it?

**Comment by rodolforg** [comment]
Known issue: TAB is not rendered properly when HarfBuzz is enabled.

**Comment by rodolforg** [comment]
Comparison: orange/yellow (whatever): Synfig Studio 1.4.0; the green one is current proposal
Captura de tela de 2021-07-18 16-52-48
Captura de tela de 2021-07-18 16-51-42

**Review by ice0 — CHANGES_REQUESTED** [review]

File: synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp [hunk 1]
```diff
@@ -692,6 +665,43 @@ Layer_Freetype::new_face(const String &newfont)
 	return true;
 }
 
+std::vector<std::string>
```
[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:676]: Do you mean to place `static` making it internal to `.cpp` or change the declaration in the `.h`?

[the remark this answers is not in the document]
Comment (ice0) [reply synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:676]: Change the declaration in the `.h`

[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:676]: Done

File: synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp [hunk 2]
```diff
@@ -214,18 +217,34 @@ struct FontMeta {
...
 };
 
+struct FaceInfo {
+	FT_Face face = nullptr;
+#if HAVE_HARFBUZZ
+	hb_font_t *font = nullptr;
+#endif
+
+	FaceInfo() = default;
+	FaceInfo(FT_Face ft_face)
```
[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp]: Done

**Comment by rodolforg** [comment]
The last forced push was to amend the following comments, due to @ice0 review:
-  Layer_FreeType: separate code for possible folders for fonts (make new method `static`)
-  Layer_FreeType: separate utf32 converter and line parser (remove the useless lambda function that I failed to implement in the next commit for FriBiDi) and solved the BiDi issue another way)
-  Layer_FreeType: add support to text shaping via HarfBuzz  (mark a construct as `explicit` and added some comments)

**Comment by rodolforg** [comment]
@FirasH could you test it?
Mixing scripts, adding texts with multiple lines, etc.?

I don't know Arabic, Hebrew, etc... So I tried to make according to the manuals, but I had to adapt to Synfig needs.

**Comment by FirasH** [comment]
> 
> 
> @FirasH could you test it?
> Mixing scripts, adding texts with multiple lines, etc.?
> 
> I don't know Arabic, Hebrew, etc... So I tried to make according to the manuals, but I had to adapt to Synfig needs.

Testing now!

**Comment by FirasH** [comment]
I cannot start the AppVeyor image on Windows, need to build it.
I'll report ASAP.

**Comment by FirasH** [comment]
1) Noticed that the word سلام is displayed with a spacing between the first and the second character as displayed here (top one is correct, made with Scribus):
arabic-text-error-1
even changing the first letter the issue is still present.
Keeping in mind Arabic is typed from right to left the problem is related to the part of text in the "middle" which is composed of two letters ل and ا

2) Noticed that the word كَلَام doesn't have the wrong spacing between first and second letter, but the "accents" are wrongly placed as displayed here (top one is correct):
arabic-text-error-2
If I remove the "accents" characters are split as before:
Screenshot_20210721_225708

**Comment by rodolforg** [comment]
Thank you, @FirasH !

I'll fix it. Luckily, this error happened after I tried to fix the case when I include an English word in an Arabic phrase like I show [a few comments above](https://github.com/synfig/synfig/pull/2232#issuecomment-882108781). I tested without that fix, and the text shapes like you showed me how it should do! :)

What about multiple lines? If a single phrase extends to next line it shapes differently if the second line would be another phrase starting with the same word? Er... So confusing. Let me show I want to say. Please consider each character-sequence as a valid word, and if a character-sequence repeats, it means they are the same word:

Case 1: one single phrase in two lines (5 words in the first line and the last 3 in the second line)
```
eeee dddd cccc bbbb aaaaa
hhhh gggg ffff.
```

Case 2: two phrases, one per line
```
.dddd cccc bbbb aaaaa
.mmmm jjjj ffff
```
My question is: does the `ffff` word could be shaped differently when comparing both cases above? It's the same word, but one is in the middle of the sentence, another (case 2) is at the sentence begining. 
Sorry for this question, but I don't know nothing about this language.

**Comment by FirasH** [comment]
> What about multiple lines? If a single phrase extends to next line it shapes differently if the second line would be another phrase starting with the same word? Er... So confusing. Let me show I want to say. Please consider each character-sequence as a valid word, and if a character-sequence repeats, it means they are the same word:
> 
> Case 1: one single phrase in two lines (5 words in the first line and the last 3 in the second line)
> 
> ```
> eeee dddd cccc bbbb aaaaa
> .hhhh gggg ffff
> ```
> 
> Case 2: two phrases, one per line
> 
> ```
> .dddd cccc bbbb aaaaa
> .mmmm jjjj ffff
> ```
> 
> My question is: does the `ffff` word could be shaped differently when comparing both cases above? It's the same word using the same characters, but one is in the middle of the sentence, another (case 2) is at the sentence begining.
> Sorry for this question, but I don't know nothing about this language.
> 
> In English, Portuguese (my mother tongue/native language), French, Spanish, German, I know that "it doesn't matter": the first character needs to be changed to "upper case", but it's another character/byte that user need to type directly, it's not automatic. In Japanese, the position in phrase doesn't matter at all.
> 
> Explaining why I'm asking it: if it doesn't matter for Arabic and other languages, I can split the text into lines first and optimize the code a bit. Otherwise I could only do it at each frame rendering (in current text renderer).

In Arabic there is no upper cases, in your example `ffff` should loook the same for both Case 1 and Case 2.

**Comment by Fahad-Alsaidi** [comment]
hi @rodolforg  
sorry for late reply... Actually I couldn't test the code, I didn't know how active freebidi & harfbuzz  while build synfig. This what I did:

> ./2-build-debug.sh all configure --with-fribidi --with-harfbuzz

**Comment by Fahad-Alsaidi** [comment]
Great work I test it, and it works fine except Firas comments

**Comment by rodolforg** [comment]
I believe I fixed it :)  @FirasH @Fahad-Alsaidi

**Comment by Fahad-Alsaidi** [comment]
> I believe I fixed it :) @FirasH @Fahad-Alsaidi

well, there is no difference see 
image

I notice one bug in many fonts in my system, the last letter  is converted to rectangular see: 
in [Dubai font](https://www.ffonts.net/Dubai.font)
image
in [Amiri Quran font](https://arabicfonts.net/fonts/amiri-quran-regular) 
image

here is a test file with fonts 
[arabic-test-file.zip](https://github.com/synfig/synfig/files/6874138/arabic-test-file.zip)

**Comment by rodolforg** [comment]
Oh my. I was so happy before… :)
Captura de tela de 2021-07-25 12-47-05
Captura de tela de 2021-07-25 12-46-55
I'll try to understand what is happening.

**Comment by rodolforg** [comment]
Hey! I think I found out the problem @Fahad-Alsaidi : synfig doesn't recognize the font file ;)
Try to write the filename (with file extension)! :D
Captura de tela de 2021-07-25 13-08-28

**Comment by Fahad-Alsaidi** [comment]
I test it again 
image

I got this error:

> error: Layer_Freetype: Unable to open font face. (err=1): Dubai_Regular.otf

how can I check that I am using the same code you have??

**Comment by rodolforg** [comment]
Are you building from source?
Test with git log and check the last commit ace87ebe9b751f53d2c93b33aff8d2c69e456e9e

**Comment by Fahad-Alsaidi** [comment]
> Are you building from source?
 I started from beginning with deleting everything, then it works. All problems gone.
Thanks for great work.

**Comment by rodolforg** [comment]
It was probably because I replaced some commits (`git push --force`) to create a clean Pull Request.
Sorry the inconvenience.
Thank you for your help.

**Review by ice0 — DISMISSED** [review]

**Comment by yoSachinkr** [comment]
Windows build doesn't work. It throughs following errors.
Screenshot 2021-07-26 182544
Screenshot 2021-07-26 182532
Screenshot 2021-07-26 182516

**Comment by ice0** [comment]
> Windows build doesn't work. It throughs following errors.

I will take a look

**Comment by rodolforg** [comment]
Commit 35733456a5744e0392a7dd556618760ad9753264 of 11/07  on master branch was fine.
Commit f32ae of 19/07 is not.

**Comment by rodolforg** [comment]
@ice0 This is the error on Catalina (CMake+Ninja): `ld: library not found for -lharfbuzz`

**Comment by ice0** [comment]
> @ice0 This is the error on Catalina (CMake+Ninja): `ld: library not found for -lharfbuzz`
> 
> Strange… it's detected by CMake itself…

Yeah, CMake `XXX_LIBRARIES` doesn't include all required flags.


> Commit [3573345](https://github.com/synfig/synfig/commit/35733456a5744e0392a7dd556618760ad9753264) of 11/07 on master branch was fine.
> Commit f32ae of 19/07 is not.

It builds fine locally, so it looks like a temporary problem with MSYS packages on AppVeyor. 
Build script doesn't update packages before installing (just to save time), so I'll check tomorrow if that helps.

**Comment by rodolforg** [comment]
> Yeah, CMake `XXX_LIBRARIES` doesn't include all required flags (I mean the paths of the libraries)
> https://cmake.org/cmake/help/latest/module/FindPkgConfig.html

So should it use `<XXX>_LINK_LIBRARIES` ? (I read it that page you linked)

**Comment by ice0** [comment]
> So should it use `<XXX>_LINK_LIBRARIES` ? (I read it that page you linked) or even `<XXX>_LDFLAGS`

I changed it to `<XXX>_LDFLAGS`. Please check PR (https://github.com/rodolforg/synfig/pull/3/)

**Comment by rodolforg** [comment]
I made it a draft again in order to wait for the CI tests. If it runs OK, I'll rebase and push again.

**Comment by morevnaproject** [comment]
Hi @rodolforg! Thank you so much for taking on this issue!

I have tested this PR and got some weird results.

I have entered a simple text into Google Translate and got the following:
screenshot_001
https://translate.google.ru/?sl=en&tl=ar&text=Synfig%20is%20an%20open-source%20animation%20software!&op=translate

Then I have copy-pasted the translated text to Synfig and got this result:
screenshot_002

Somehow the words are in different order.

I am on Linux Mint 19, using Russian locale. Same result when I start Synfig with English (LANG=C) locale.

Attaching my sample file - 
[text-arabic-animatied-xy-slow.sifz.zip](https://github.com/synfig/synfig/files/6921305/text-arabic-animatied-xy-slow.sifz.zip)

@FirasH @Fahad-Alsaidi can you please test this file on your system and give screenshots how it looks? I am not familiar with Arabic language, maybe this is normal... ^__^"

**Comment by Fahad-Alsaidi** [comment]
Hi @morevnaproject 
to get full RTL support, you will need to enforce RTL direction flag to the text.
See this screen shot from libreoffice when I set LTR flag:
image
here when set RTL flag
image

The simplest solution is to implement a radio button for setting RTL

**Comment by morevnaproject** [comment]
@Fahad-Alsaidi I have noticed that in your examples all Arabic words are still remain in the same order, independently of RTL/LTR flag  (I have highlighted that part below) -

screenshot_003

The problem is that in my case Arabic words displayed in different order. Also, you can compare with the same text in Params panel (highlighted on screenshot below) - 
screenshot_002

**Comment by rodolforg** [comment]
> Somehow the words are in different order.
@morevnaproject 
It seems a font issue. Use an Arabic font :)

Anyway, it indeed has an error:
Captura de tela de 2021-08-03 03-38-07

GTK renders texts wrong here, as it supposes it's LTR. I don't know if we can fix it. But our text rendering is possible to fix by using @Fahad-Alsaidi suggestion: adding an extra layer parameter: LTR/RTL (an enum? or a boolean?)

**Comment by Fahad-Alsaidi** [comment]
this mine 
image

**Comment by ice0** [comment]
> GTK renders texts wrong here, as it supposes it's LTR. I don't know if we can fix it. But our text rendering is possible to fix by using @Fahad-Alsaidi suggestion: adding an extra layer parameter: LTR/RTL (an enum? or a boolean?)

As discussed here (https://github.com/microsoft/vscode/issues/86667#issuecomment-631025893):

> I offer support for Unicode RTL character (200F) the editor. It ‏works in some programs like Telegram and GTK applications.

**Comment by rodolforg** [comment]
@Fahad-Alsaidi Is the only problem "Synfig" word position, right? It should be at most-right position.

**Comment by rodolforg** [comment]
> I offer support for Unicode RTL character (200F) the editor. It ‏works in some programs like Telegram and GTK applications.

It indeed works fine if someone type/paste this Unicode mark ("character")

**Comment by ice0** [comment]
> > I offer support for Unicode RTL character (200F) the editor. It ‏works in some programs like Telegram and GTK applications.
> 
> It indeed works fine if someone type/paste this Unicode mark ("character")

I was hoping that the OS will insert it itself if it understands that the text is written from right to left.  😅

**Comment by rodolforg** [comment]
> I was hoping that the OS will insert it itself if it understands that the text is written from right to left. sweat_smile

I hoped it too. But I didn't think so: GTK+Pango seems to have a pretty good support for Arabic, but we can see it renders wrong the Konstantin's text in this comment: https://github.com/synfig/synfig/pull/2232#issuecomment-891546996 . Check Google Translator text versus Layer Panel versus Parameter Panel. The last two cases are pure GTK code and it renders in wrong order.

**Comment by khaledhosny** [comment]
Paragraph direction can’t be automatically set, you can default to LTR or RTL based on the locale but 1) the setting must be stored in the file (otherwise opening the same file on a system with a different locale can give different results) 2) the user mut have a way to override it. You can pass paragraph direction to FriBiDi. I suggest either using [Raqm](https://github.com/HOST-Oman/libraqm) instead of wrapping HarfBuzz and FriBiDi yourself, or study its code for some pointers.

**Comment by khaledhosny** [comment]
There is a heuristic that sets the paragraph direction based on the first character with a strong BiDi direction, but it fails in cases like this one. The RLM control character is one solution, but it is an invisible character and most users can‘t type it, so inserting it is rather inconvenient.

**Comment by rodolforg** [comment]
So we can't escape from adding a new parameter: text direction LTR/RTL, right?

**Comment by khaledhosny** [comment]
> So we can't escape from adding a new parameter: text direction LTR/RTL, right?

I think so.

**Comment by morevnaproject** [comment]
> @morevnaproject
> It seems a font issue. Use an Arabic font :)

I tested it with "Noto Naksh Arabic" (was in my system) and also downloaded ["Amiri" font from Google Fonts](https://fonts.google.com/specimen/Amiri?subset=arabic) - same result... :-/
screenshot_003



> Anyway, it indeed has an error:
> Captura de tela de 2021-08-03 03-38-07

Is this screenshot made on Linux?

**Comment by rodolforg** [comment]
> Is this screenshot made on Linux?

Yes. I almost never have access to a Windows machine.

**Comment by khaledhosny** [comment]
Noto fonts are split into one script per font family, do Noto Naskh Arabic supports Arabic characters only, and it appears that synfig does not support font fallback. Fallback is very important for text layout since any given font will only support a subset of Unicode (it is not even technically possible for one font to support all currently defined Unicode characters, not with the existing font formats).

Since synfig already depends on GTK, has Pango been considered? It implements solutions for all of the issues discussed here and more.

**Comment by rodolforg** [comment]
> I suggest either using [Raqm](https://github.com/HOST-Oman/libraqm) instead of wrapping HarfBuzz and FriBiDi yourself, or study its code for some pointers.


> Since synfig already depends on GTK, has Pango been considered? It implements solutions for all of the issues discussed here and more.

Synfig itself does not depend on GTK, but Synfig Studio (the GUI) does.

Somehow some day I found the `raqm` library. It does provide the text-shaping solution, but only as a bitmap text directly (as long as I saw the code): we have to pass the FreeFont font face with font/character size set. That is what makes text layer currently sooo slow. I already ported text layer to current synfig render engine, in a vector way and it's quite fast (as expected).

Pango has the same problem raqm, with another downside: I can't reproduce all current features of text layer.

I'll check how can I see it's an invalid glyph in order to (try to) use a fallback font face.

**Comment by rodolforg** [comment]
> I tested it with "Noto Naksh Arabic" (was in my system) and also downloaded ["Amiri" font from Google Fonts](https://fonts.google.com/specimen/Amiri?subset=arabic) - same result... :-/
> 
> screenshot_003

@morevnaproject do you have FriBiDi and HarfBuzz installed?

**Comment by khaledhosny** [comment]
Raqm does not do any rendering, it returns glyph IDs and positions. It needs size set on the FT_Face to be able to return glyph positions, but you can set it to the fonts UPEM if you don’t want it to do any scaling.

**Comment by rodolforg** [comment]
Oh, thank you for correcting me. I looked an example code and it forced a size.
Anyway, for now I don't see any implementation difference between my code and raqm. As raqm isn't distributed in Debian, I'll try to keep away from this library for now.

I added the text direction parameter (Auto, LTR, RTL). I'll think now how to make the font fallback :/ I didn't want to add a new parameter nor make the "font family" a list :(

**Comment by morevnaproject** [comment]
> @morevnaproject do you have FriBiDi and HarfBuzz installed?

Woops, I found that `fribidi-dev` is not installed.
We should add additional check to build system (I have used Autotools). ^__^

Rebuilding now...

**Comment by morevnaproject** [comment]
@rodolforg I have made attempt to build latest commit with fribidi-dev installed (package version 0.19.7-2) and got the following error:
```
Making all in lyr_freetype
make[4]: Entering directory '/home/konstantin/sources/synfig/_production/synfig-core/src/modules/lyr_freetype'
  CXX      liblyr_freetype_la-lyr_freetype.lo
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp: In static member function 'static std::vector<std::vector<Layer_Freetype::TextSpan> > Layer_Freetype::fetch_text_lines(const string&, int)':
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:15: error: 'FriBidiBracketType' was not declared in this scope
   std::vector<FriBidiBracketType> bracket_types(line_size);
               ^~~~~~~~~~~~~~~~~~
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:15: note: suggested alternative: 'FriBidiParType'
   std::vector<FriBidiBracketType> bracket_types(line_size);
               ^~~~~~~~~~~~~~~~~~
               FriBidiParType
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:33: error: template argument 1 is invalid
   std::vector<FriBidiBracketType> bracket_types(line_size);
                                 ^
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:33: error: template argument 2 is invalid
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1368:102: error: request for member 'data' in 'bracket_types', which is of non-class type 'int'
 ribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
                                                                     ^~~~
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1368:33: error: 'fribidi_get_par_embedding_levels_ex' was not declared in this scope
   FriBidiLevel fribidi_result = fribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1368:33: note: suggested alternative: 'fribidi_get_par_embedding_levels'
   FriBidiLevel fribidi_result = fribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                 fribidi_get_par_embedding_levels
Makefile:613: recipe for target 'liblyr_freetype_la-lyr_freetype.lo' failed
make[4]: *** [liblyr_freetype_la-lyr_freetype.lo] Error 1
make[4]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core/src/modules/lyr_freetype'
Makefile:568: recipe for target 'all-recursive' failed
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core/src/modules'
Makefile:471: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core/src'
Makefile:693: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core'
Makefile:561: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by rodolforg** [comment]
> > @morevnaproject do you have FriBiDi and HarfBuzz installed?
> 
> Woops, I found that `fribidi-dev` is not installed.
> We should add additional check to build system (I have used Autotools). ^__^

But it does have a check for FriBiDi o.O

**Comment by morevnaproject** [comment]
> It isn't mandatory though.

I guess this is why it gave me different text rendering result. I think we should make it mandatory. ^__^

**Comment by rodolforg** [comment]
> with fribidi-dev installed (package version 0.19.7-2)

Ouch. The API changed for 1.0.0 version.
https://github.com/fribidi/fribidi/releases?after=v1.0.1

I'll place a version detection

**Comment by morevnaproject** [comment]
I think we should make fribidi  a required dependency - i.e. the configuration process should fail if no fribidi headers found.

Reason for that:
Synfig should give reproducible output from the same files.

Imagine one user creates animation with arabic text. He has a version of Synfig compiled with fribidi, so everything looks good for him.
Then he sends the animation file to another user (or maybe to cloud renderfarm? ^__^) and this second user has a version of Synfig without fribidi. So, he gets a different output and that can lead to confusions.

So I think we should make fribidi a required package to avoid such situations. ^__^

**Comment by rodolforg** [comment]
>  Synfig should give reproducible output from the same files.

So we should require Harfbuzz too, right?

**Comment by morevnaproject** [comment]
> So we should require Harfbuzz too, right?

Yes, Harfbuzz should be a mandatory requirement too. ^__^

**Comment by morevnaproject** [comment]
> Yes, Harfbuzz should be a mandatory requirement too. ^__^

Note: Being said that, I am not sure which effect gives compiling without Harfbuss. I have made attempt to remove `hafbuzz-dev` package from my system , but that requested to remove many other useful packages. So I haven't did that. But I pretty much sure that compiling without Harfbuzz also makes RTL text look different. ^__^

**Comment by rodolforg** [comment]
FriBiDi handles bidirectional text. 
HarfBuzz focus on text/character shaping. Not exactly on mixing text direction.

So, without harfbuzz is for those who doesn't need to handle Arabic, Hebrew, etc., i.e., as it is now.

**Comment by morevnaproject** [comment]
> So, without harfbuzz is for those who doesn't need to handle Arabic, Hebrew, etc., i.e., as it is now.

I see. So, I still think Harfbuzz should be a mandatory requirement - to avoid different output id different conditions. We can't predict if user wants bi-directional/shaping text or not. ^__^

**Comment by rodolforg** [comment]
@morevnaproject  Last commit I made harfbuzz required by default. If it is required, FriBiDi is automatically required too.

I still don't provide any support for fallback fonts :(

I don't have any idea about FriBiDi 0.18.x API so I can't implement it, and Ubuntu 16.04 building fails

**Comment by rodolforg** [comment]
> @rodolforg I have made attempt to build latest commit with fribidi-dev installed (package version 0.19.7-2) and got the following error:
> 
> ```
> Making all in lyr_freetype
> make[4]: Entering directory '/home/konstantin/sources/synfig/_production/synfig-core/src/modules/lyr_freetype'
>   CXX      liblyr_freetype_la-lyr_freetype.lo
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp: In static member function 'static std::vector<std::vector<Layer_Freetype::TextSpan> > Layer_Freetype::fetch_text_lines(const string&, int)':
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:15: error: 'FriBidiBracketType' was not declared in this scope
>    std::vector<FriBidiBracketType> bracket_types(line_size);
>                ^~~~~~~~~~~~~~~~~~
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:15: note: suggested alternative: 'FriBidiParType'
>    std::vector<FriBidiBracketType> bracket_types(line_size);
>                ^~~~~~~~~~~~~~~~~~
>                FriBidiParType
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:33: error: template argument 1 is invalid
>    std::vector<FriBidiBracketType> bracket_types(line_size);
>                                  ^
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1367:33: error: template argument 2 is invalid
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1368:102: error: request for member 'data' in 'bracket_types', which is of non-class type 'int'
>  ribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
>                                                                      ^~~~
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1368:33: error: 'fribidi_get_par_embedding_levels_ex' was not declared in this scope
>    FriBidiLevel fribidi_result = fribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
>                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/konstantin/sources/synfig/synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp:1368:33: note: suggested alternative: 'fribidi_get_par_embedding_levels'
>    FriBidiLevel fribidi_result = fribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
>                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>                                  fribidi_get_par_embedding_levels
> Makefile:613: recipe for target 'liblyr_freetype_la-lyr_freetype.lo' failed
> make[4]: *** [liblyr_freetype_la-lyr_freetype.lo] Error 1
> make[4]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core/src/modules/lyr_freetype'
> Makefile:568: recipe for target 'all-recursive' failed
> make[3]: *** [all-recursive] Error 1
> make[3]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core/src/modules'
> Makefile:471: recipe for target 'all-recursive' failed
> make[2]: *** [all-recursive] Error 1
> make[2]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core/src'
> Makefile:693: recipe for target 'all-recursive' failed
> make[1]: *** [all-recursive] Error 1
> make[1]: Leaving directory '/home/konstantin/sources/synfig/_production/synfig-core'
> Makefile:561: recipe for target 'all' failed
> make: *** [all] Error 2
> ```

@morevnaproject I took a look at FriBiDi 0.19.7 source code.
You should replace
```c++
		std::vector<FriBidiBracketType> bracket_types(line_size);
		FriBidiLevel fribidi_result = fribidi_get_par_embedding_levels_ex(bidi_types.data(), bracket_types.data(), line_size,
															  &base_dir, bidi_levels.data());
```
with
```
		FriBidiLevel fribidi_result = fribidi_get_par_embedding_levels(bidi_types.data(), line_size,
															  &base_dir, bidi_levels.data());
```

**Comment by rodolforg** [comment]
We could use remove the FriBiDi > 1.0.0 requirement and select code with `#if FRIBIDI_MAJOR_VERSION >= 1` and implement my proposed solution to your compilation problem with correspondent `#else`.

**Comment by morevnaproject** [comment]
> We could remove the FriBiDi > 1.0.0 requirement and select code with `#if FRIBIDI_MAJOR_VERSION >= 1` and implement my proposed solution to your compilation problem with correspondent `#else`.

That would be great! I need to check the code you suggested first. Maybe there are other compilation issues...

**Comment by rodolforg** [comment]
@ice0 could you please review https://github.com/synfig/synfig/pull/2232/commits/8e4375f4835e5d81c58cb330160d98cf1a2a7c20 ?

**Comment by rodolforg** [comment]
> That would be great! I need to check the code you suggested first. Maybe there are other compilation issues...

@morevnaproject sorry I read the first phrase and then I pushed XD

**Comment by rodolforg** [comment]
If the intention is to Animation document be reproducible I'm thinking about a new field: font fallback.
Otherwise each system could have its own fonts and the fallback could be different, rendering differently...
Argh.

**Comment by morevnaproject** [comment]
@rodolforg Also, can you please add `fribidi-dev` package to this list? This package seems to be not present by default...
https://github.com/synfig/synfig/blob/2cbce018087fe63227afa099760ce83ecb4005f3/1-setup-linux-native.sh#L200-L232

**Comment by morevnaproject** [comment]
> @morevnaproject sorry I read the first phrase and then I pushed XD

Wohooo! Now I have correct output for Arabic text!
screenshot_006
Thank you!

**Comment by morevnaproject** [comment]
> If the intention is to Animation document be reproducible I'm thinking about a new field: font fallback.
> Otherwise each system could have its own fonts and the fallback could be different, rendering differently...
> Argh.

I think this is a responsibility of user to ship required fonts with his animation project.
In our practice, when sharing animation project we always put all font files into special subdirectory of the project. And in Synfig files we set relative paths to font files, so Synfig can easily find them when project is opened on any machine.

In fact the possibility to specify path to font files in Synfig is a great feature, which allows to solve problem of missing fonts in system.

So I think there is no need to bother with font replacement.

I think this PR is ready for merging! ^__^

**Review by morevnaproject — APPROVED** [review]

**Comment by rodolforg** [comment]
> @rodolforg Also, can you please add `fribidi-dev` package to this list? This package seems to be not present by default...

Done. I didn't know if it was needed by ALT Linux and what would be the package name.

**Comment by morevnaproject** [comment]
> I didn't know if it was needed by ALT Linux and what would be the package name.

For Alt Linux the package name seems to be `libfribidi-devel` - http://sisyphus.ru/en/srpm/Sisyphus/fribidi/repocop

**Comment by rodolforg** [comment]
Trying to fix CMake build for MacOS I broke...

> > I didn't know if it was needed by ALT Linux and what would be the package name.
> 
> For Alt Linux the package name seems to be `libfribidi-devel` - http://sisyphus.ru/en/srpm/Sisyphus/fribidi/repocop

@morevnaproject done :)

**Comment by morevnaproject** [comment]
@rodolforg Can you rebase, and resolve conflicts, please? ^__^

**Comment by rodolforg** [comment]
> @rodolforg Can you rebase, and resolve conflicts, please? ^__^

I'll do it now! :)

**Comment by rodolforg** [comment]
@morevnaproject  done

**Review by morevnaproject — APPROVED** [review]

**Comment by morevnaproject** [comment]
@ice0 Can we merge this? ^__^

**Comment by ice0** [comment]
> @ice0 Can we merge this? ^__^

@morevnaproject Sure. 
@rodolforg Can you rebase please?

**Comment by rodolforg** [comment]
@ice0 done

**Review by ice0 — APPROVED** [review]

**Comment by ice0** [comment]
Merged. Thank you!

**Comment by morevnaproject** [comment]
Woohooo! :tada: :tada: :tada:

**Comment by morevnaproject** [comment]
@rodolforg Now eagerly waiting a PR with optimized (cobra-enabled) text layer. ^__^

**Comment by Fahad-Alsaidi** [comment]
I would like to thank @rodolforg & @morevnaproject & @ice0  for supporting RTL in synfig. You did a great job, thanks 🌹🌹

**Comment by rodolforg** [comment]
> @rodolforg Now eagerly waiting a PR with optimized (cobra-enabled) text layer. ^__^

I'm working to make a clean version of what I wrote before HarfBuzz :) It still works, but I need to restore text alignment ("orientation" parameter).

@Fahad-Alsaidi 
Sadly we still don't provide a fallback-font for languages not supported by the chosen font.

**Pull request merged**